Assets/BossRoom/Scripts/Shared/Net/UnityServices/Lobbies/LobbyAsyncRequests.cs

```diff
@@ -0,0 +1,339 @@
+                // the actual rate limits are tracked via the RateLimitCooldown objects defined above
+                m_SlowUpdate.Subscribe(UpdateLobby, 1.5f);
+            }
+        }
+
+        public void EndTracking()
+        {
+            if (m_isTracking)
+            {
+                m_SlowUpdate.Unsubscribe(UpdateLobby);
+                m_isTracking = false;
+                m_lastKnownLobby = null;
+                m_heartbeatTime = 0;
+            }
+        }
+
+        private void UpdateLobby(float unused)
+        {
+            RetrieveLobbyAsync(m_LocalLobby.LobbyID, OnSuccess, null);
+
+            void OnSuccess(Lobby lobby)
+            {
+                m_lastKnownLobby = lobby;
+            }
+        }
+
+        #endregion
+
+        #region Lobby API calls are rate limited, and some other operations might want an alert when the rate limits have passed.
+
+        // Note that some APIs limit to 1 call per N seconds, while others limit to M calls per N seconds. We'll treat all APIs as though they limited to 1 call per N seconds.
+        // Also, this is seralized, so don't reorder the values unless you know what that will affect.
+        public enum RequestType
+        {
+            Query = 0,
+            Join,
+            QuickJoin,
+            Host
+        }
+
+        public RateLimitCooldown GetRateLimit(RequestType type)
+        {
+            if (type == RequestType.Join)
+            {
+                return m_RateLimitJoin;
+            }
+            else if (type == RequestType.QuickJoin)
+            {
+                return m_RateLimitQuickJoin;
+            }
+            else if (type == RequestType.Host)
+            {
+                return m_RateLimitHost;
+            }
+
+            return m_RateLimitQuery;
+        }
+
+        private RateLimitCooldown m_RateLimitQuery; // Used for both the lobby list UI and the in-lobby updating. In the latter case, updates can be cached.
+        private RateLimitCooldown m_RateLimitJoin;
+        private RateLimitCooldown m_RateLimitQuickJoin;
+        private RateLimitCooldown m_RateLimitHost;
+
+        #endregion
+
+        /// <summary>
+        /// Attempt to create a new lobby and then join it.
+        /// </summary>
+        public void CreateLobbyAsync(string lobbyName, int maxPlayers, bool isPrivate, OnlineMode onlineMode, string ip, int port, Action<Lobby> onSuccess, Action onFailure)
+        {
+            if (!m_RateLimitHost.CanCall())
+            {
+                onFailure?.Invoke();
+                UnityEngine.Debug.LogWarning("Create Lobby hit the rate limit.");
+                return;
+            }
+
+            string uasId = AuthenticationService.Instance.PlayerId;
+
+            var initialLobbyData = new Dictionary<string, DataObject>()
+            {
+                {"OnlineMode", new DataObject(DataObject.VisibilityOptions.Public, ((int)onlineMode).ToString())},

```

Comment:
what happens if I want to rename this to "BossRoomOnlineMode"?
I'll need to update it everywhere else it's used.
This Dictionary should be a struct really, here we're sort of writing some weird more complex json...
Can you send direct json instead of a Dictionary? Cause you could use JsonUtility for this and manage your own struct, this way you'd be able to have compile time typing in the data you send.
Sending big untyped string to object Dictionaries like this really isn't good practices when dealing with REST APIs client side.
Or if no struct possibility (eurk), at least create consts for these strings.

Reply:
weird-mode JSON sounds like normal web development to me :P

Reply:
And this isn't web dev, this REST services dev. Ain't no web page here

Reply:
I was only joking. So a class called `CreateLobbyOptions` is a part of the lobby sdk and it requires us to pass it a Dictionary<stirng, DataObject> - it basically insists on this weird-form JSON thing.

```
namespace Unity.Services.Lobbies
{
    /// <summary>
    /// Optional parameter class for Lobby creation requests.
    /// </summary>
    public class CreateLobbyOptions 
    {
        /// <summary>
        /// Indicates whether or not the lobby is publicly visible and will show up in query results.
        /// If the lobby is not publicly visible, the creator can share the LobbyCode with other users who can use it to join this lobby.
        /// If left as null, a default value will be used.
        /// </summary>
        public bool? IsPrivate { get; set; }

        /// <summary>
        /// Information about a specific player creating the lobby.
        /// </summary>
        public Player Player { get; set; }

        /// <summary>
        /// Custom game-specific properties that apply to the lobby (e.g. mapName or gameType).
        /// </summary>
        public Dictionary<string, DataObject> Data { get; set; }
    }
}
```

Reply:
We can clean it up in a struct that can convert itself into a dictionary for the service api, but it's a minor tweak that I'd rather do in the next PR